examples/generative/stylegan.py

```diff
@@ -0,0 +1,736 @@
+TRAIN_STEP_RATIO = {k: BATCH_SIZE[2] / v for k, v in BATCH_SIZE.items()}
+
+
+def load(res, image_file):
+    image = tf.io.read_file(image_file)
+    image = tf.image.decode_jpeg(image)
+    image = tf.image.resize(
+        image, [res, res], method=tf.image.ResizeMethod.NEAREST_NEIGHBOR
+    )
+    image = tf.cast(image, tf.float32)
+    image = (image / 127.5) - 1
+    return image
+
+
+def create_dataset_(data_path, res):
+    res_log2 = int(np.log2(res))
+    BUFFER_SIZE = 100
+    images = glob(data_path)
+    random.shuffle(images)
+    batch_size = BATCH_SIZE[res_log2]
+    tf_list = tf.data.Dataset.from_tensor_slices(images)
+    dataset = tf_list.map(
+        partial(load, res), num_parallel_calls=tf.data.experimental.AUTOTUNE
+    )
+    dataset = (
+        dataset.shuffle(BUFFER_SIZE).batch(batch_size, drop_remainder=True).repeat()
+    )
+    return dataset
+
+
+create_dataset = partial(create_dataset_, DATA_PATH)
+
+
+"""
+### Utility function to display images after each epoch
+"""
+
+
+def plot_images(images, log2_res, fname=""):
+    scales = {2: 0.5, 3: 1, 4: 2, 5: 3, 6: 4, 7: 5, 8: 6, 9: 7, 10: 8}
+    scale = 0.5 * scales[log2_res]
+
+    grid_col = min(images.shape[0], int(16 // scale))
+    grid_row = 1
+
+    f, axarr = plt.subplots(
+        grid_row, grid_col, figsize=(grid_col * scale, grid_row * scale)
+    )
+
+    for row in range(grid_row):
+        ax = axarr if grid_row == 1 else axarr[row]
+        for col in range(grid_col):
+            ax[col].imshow(images[row * grid_col + col])
+            ax[col].axis("off")
+    plt.show()
+    if fname:
+        f.savefig(fname)
+
+
+"""
+## Custom Layers
+The following are building blocks that will be used to construct the generators and
+discriminators.
+"""
+
+
+def fade_in(alpha, a, b):
+
+    return alpha * a + (1.0 - alpha) * b
+
+
+def wasserstein_loss(y_true, y_pred):
+
+    return -tf.reduce_mean(y_true * y_pred)
+
+
+class PixelNorm(Layer):
+    def __init__(self, epsilon=1e-8):
+        super(PixelNorm, self).__init__()
+        self.epsilon = epsilon
+
+    def call(self, input_tensor):
+        return input_tensor / tf.math.sqrt(
+            tf.reduce_mean(input_tensor ** 2, axis=-1, keepdims=True) + self.epsilon
+        )
+
+
+class MinibatchStd(Layer):
+    def __init__(self, group_size=4, epsilon=1e-8):
+
+        super(MinibatchStd, self).__init__()
+        self.epsilon = epsilon
+        self.group_size = group_size
+
+    def call(self, input_tensor):
+
+        n, h, w, c = input_tensor.shape
+        x = tf.reshape(input_tensor, [self.group_size, -1, h, w, c])
+        group_mean, group_var = tf.nn.moments(x, axes=(0), keepdims=False)
+        group_std = tf.sqrt(group_var + self.epsilon)
+        avg_std = tf.reduce_mean(group_std, axis=[1, 2, 3], keepdims=True)
+        x = tf.tile(avg_std, [self.group_size, h, w, 1])
+
+        return tf.concat([input_tensor, x], axis=-1)
+
+
+class ConvBlock(Layer):
```

Comment:
The name should be more descriptive. `ScaledConv`?

Reply:
Renamed to `EqualizedConv` for equalized learning rate.